### ISSUE TITLE
[iOS] bug report - can’t get past login

### ISSUE THREAD

**Issue by 3y3lash** [zulip/zulip-mobile#4149]
iOS on iPhone 7 running 12.2, can’t get past login dialog. After entering correct server, account and password the app keeps redirecting to back to login. Infinite loop.
Not sure where else to report. 
Same login details working fine on multiple desktop versions.

**Comment by gnprice** [comment]
That's no good! Thanks for the report.

A few questions to help us diagnose and resolve this:
* Is the server you're using at zulipchat.com, or is it one that you / your organization run yourself? If the latter, what version of Zulip Server is it running?
* What version of the app are you using? (If you installed it just recently from the App Store, that's enough to answer this question.)
* Can you post a screenshot of what it looks like at the step just before the redirect happens?

**Comment by agrawal-d** [comment]
Does not repo on an Android device.

**Comment by gnprice** [comment]
> Does not repro on an Android device.

My guess is that this will turn out to be independent of OS, but instead depend on some interaction with the way that authentication works in the organization @3y3lash is using. But hopefully we'll learn more.

**Comment by 3y3lash** [comment]
Hey, thanks for your response - sorry for the late reply!
1) it’s on our own server running the latest version
2) I’ve been using the latest version of the app- I uninstalled and reinstalled a couple of times to check. 
3) here:
![C65FA4C0-4AA5-4FB8-8898-50D79472C00F](https://user-images.githubusercontent.com/66748310/84501122-37e4fa80-acb6-11ea-803c-e79d44033d46.png)
4) thanks!

**Comment by gnprice** [comment]
@3y3lash Thanks, that's helpful!

What do you do to log in after that screen? I imagine you touch the button that has your account email and server URL -- then what does the next screen look like? And what does the screen look like that you see just before taking the step that gets you redirected back to login?

**Comment by 3y3lash** [comment]
Hi - yep, I tap the button with my account and server and it flashes through the same sequence of logins (too fast to properly see or screenshot) eg:
![250EE67E-2DF7-4E93-AE5D-9CA4EB6CA58D](https://user-images.githubusercontent.com/66748310/84546678-20336380-ad02-11ea-9c54-1d14633e2a28.png)

Until I land back here:
![1384A19F-9837-409B-961A-7275A62D45E9](https://user-images.githubusercontent.com/66748310/84546746-4527d680-ad02-11ea-9e53-75e39dd2af9f.png)

**Comment by 3y3lash** [comment]
When I click “log in with password” it brings me back to password log in, then I enter my password and the cycle repeats.

**Comment by gnprice** [comment]
Very helpful, thanks!

Based on that sequence, and from studying this code, here's what appears to be happening:
* You're doing a password login, so using `PasswordAuthScreen` in the code.
* The way that screen finishes is here:
```js
      const fetchedKey = await api.fetchApiKey(partialAuth, email, password);
      this.setState({ progress: false });
      dispatch(loginSuccess(partialAuth.realm, fetchedKey.email, fetchedKey.api_key));
```
The `dispatch(loginSuccess(…))` is what causes that screen to go away -- any failures within that screen just cause an error message (like "Wrong email or password. Try again.") on the same screen.
* In particular, we're successfully completing the request that sends your email and password to the Zulip server and retrieves your API key (that's the `api.fetchApiKey`.)
* After that you should see the green loading screen, at least briefly -- that's the next thing we show, thanks to the handling of `LOGIN_SUCCESS` in `navReducers.js`.
* Meanwhile, we start fetching your Zulip data from the server. (The chain that causes this is a bit complicated: the handling of `LOGIN_SUCCESS` in `sessionReducers.js`, which sets `needsInitialFetch`, which causes `AppDataFetcher` to invoke `doInitialFetch`, which does the work.)
* In making *that* request, we have the following logic:
```js
  } catch (e) {
    // This should only happen on a 4xx HTTP status, which should only
    // happen when `auth` is no longer valid.  No use retrying; just log out.
    dispatch(logout());
    return;
  }
```
So if the server rejects the request saying there's a client-side error in it (that's what a 4xx HTTP status means), then we conclude that our credentials aren't going to work and take you to the login screen.
* So it seems like that is probably what's happening here.

In short, I think what's probably happening is:
* we successfully get your API key from the Zulip server, using your email and password;
* but then when we try to use that API key to get your data, the server rejects the request, saying it's invalid.

**Comment by gnprice** [comment]
@3y3lash Are you one of the people who run your Zulip server? I think the next step for debugging this is to see what the server's logs say when you try to log in.

Specifically I am curious about the contents of `/var/log/zulip/server.log` and `/var/log/zulip/errors.log` for that sequence of requests:
* In `server.log` there should be a line for a request mentioning `fetch_api_key` (that's the one where we use the password to get the API key), and then shortly after it there should be one mentioning `register` (that's the main initial one where we use your API key to get your Zulip data.)
* I'm curious about those lines, and other requests around them from the same client. I'm especially curious about the HTTP status (it comes after the word "POST") of the `register` request.
* There may also be some entries in `errors.log` related to those requests -- if there are, those are likely to be extremely helpful.

If you (or your server admins) would be more comfortable sharing the logs off this public thread, I'd be glad to see them in PMs on chat.zulip.org, or by email to support@zulipchat.com.

**Comment by 3y3lash** [comment]
Hey again - I’m not the person running the server, but I’ll pass this info on. Thank you!

**Comment by 3y3lash** [comment]
Hey there, 
An update from the person running the server:

> This is what appeared in server.log during your login attempt:
2020-06-24 11:47:57.277 INFO [zr] x.x.x.x   POST    200 134ms (db: 31ms/4q) (+start: 10ms) /api/v1/fetch_api_key (user28@zulip.example.com via ZulipMobile)
2020-06-24 11:47:57.485 INFO [zr] x.x.x.x   POST    401   4ms (+start: 6ms) /api/v1/users/me/presence (unauth via ?)
2020-06-24 11:47:57.498 INFO [zr] x.x.x.x   POST    401   3ms /api/v1/register (unauth via ?)
2020-06-24 11:47:57.516 INFO [zr] x.x.x.x   POST    401   2ms /api/v1/users/me/presence (unauth via ?)
2020-06-24 11:47:57.562 INFO [zr] x.x.x.x   GET     200  67ms (db: 11ms/2q) (+start: 5ms) /api/v1/server_settings (unauth via ZulipMobile)
2020-06-24 11:47:57.683 INFO [zr] x.x.x.x   POST    401   3ms (+start: 9ms) /api/v1/register (unauth via ?)
2020-06-24 11:47:58.386 INFO [] Tornado 9993   1.4% busy over the past 59.0 seconds
2020-06-24 11:47:59.345 INFO [zr] x.x.x.x   GET     200  60ms (db: 9ms/2q) (+start: 11ms) /api/v1/server_settings (unauth via ZulipMobile)
Nothing appeared in the error.log. I replaced the URL to our zulip instance with zulip.example.com and replaced the IP address with x.x.x.x.

**Comment by gnprice** [comment]
That's helpful, thanks!

Reading that log, what I see is:
* There's a `/fetch_api_key` request that succeeds (status 200) and looks normal.
* Then a `/register` request that fails, with a 4xx status! That matches the story I described above, based on the code. In particular it would explain exactly the symptoms you're seeing.
* Specifically it is failing with status 401, "Unauthorized". That means the server rejected the credentials we presented. (Relatedly the logging says "unauth" instead of "user28@zulip.example.com", meaning the request isn't recognized as a logged-in user.)
* So the next question is, why are those credentials being rejected? We did after all just get the API key. The credentials are just email plus API key.

I have a couple of questions for you which might help narrow the issue down:
* In the email address you use to log in, is there anything interesting like non-ASCII characters?
* On the very first screen where you enter the server URL, is the URL exactly the same as the one you see when using Zulip in a browser? If there are somehow two different URLs the server can be reached on, and the request comes on the URL that it's not expecting, I think that might produce an error that looks like this (based on a few minutes just now rereading the server's auth code.)

**Comment by gnprice** [comment]
There are also a couple of things we should probably do to make situations like these easier to debug, either for you or if they recur for someone else in the future:
* As I said above, if we get a 4xx error when trying to load your data from the server, we conclude that the credentials we have for you aren't going to work and take you to the login screen.
* When we do that, we should show an error message saying that's what happened.
* That error message should probably include the error message we get from the server -- I see in the relevant server code that there's a range of specific error messages it provides. These may often not be all that helpful for the person looking at them... but if it comes to a bug report like this one, a screenshot of such a message could be super helpful for diagnosing the problem.
* Possibly we should also notice if this happens on the very first time we try to load your data right after you've logged in. An auth failure can be perfectly normal if you've changed your API key, or deactivated your account, etc. -- but an auth failure immediately after login is almost certainly a bug somewhere.

**Comment by timabbott** [comment]
What version is the server running in this case?  If the server is using an early version of the `email_address_visibility` feature, it's possible that could explain this symptom of the server returning an invalid credential (because it's returning `user15@zulip.example.com` as the username).  It looks like we had that bug and accidentally fixed in zulip/zulip f00a852697811f647591bd5faa0d7a8ad9c0df6e, which was soon after the 2.1.0 release, and never backported, likely because we didn't noticed we'd fixed it.

I suspect upgrading the server to master (or the soon to be tagged 3.0-rc1) will resolve this.

**Comment by chrisbobbe** [comment]
Hi, @3y3lash, thanks for the report! I'm closing this, as it was likely due to a server bug that has been fixed. But if you're having this issue, please reply and we'll reopen and investigate.

**Comment by 3y3lash** [comment]
Hey Chris, thanks for getting back to me! unfortunately the problem persists :(
It still crashes on opening, even after upgrading. I looked into it further a few weeks ago and it seems like a lot of other users have been having this issue too.

**Comment by chrisbobbe** [comment]
> It still crashes on opening

That's no good! Thanks for letting me know!

When you say it crashes on opening, it sounds like the symptom might be different from what's been described in this thread. Still quite bad, but possibly it should get its own, new issue. In particular, are you having trouble immediately upon opening the app, or a little bit later, like when you try to log into your account? And is it a crash where the whole app quits, or its screen turns blank, or something else? That'll help us diagnose the issue.

If the symptoms are exactly as described in this thread, that'd be helpful to know too—but I'd be a little surprised, given what Tim said about a server issue.

**Comment by chrisbobbe** [comment]
> even after upgrading

Do you mean upgrading the mobile app, or the server, or both? It'd be helpful to know what versions of both you're currently running (when you see the issue).

**Comment by chrisbobbe** [comment]
> I looked into it further a few weeks ago

Do you think the same problem might have persisted continuously since June of last year, or might there have been a period where things were working OK?

**Comment by chrisbobbe** [comment]
> it seems like a lot of other users have been having this issue too.

Are these users in your same org, connecting to the same server, or have you heard from others who are connecting to different servers?

**Comment by 3y3lash** [comment]
So, I tap on Signal icon, the blue signal splash screen with the signal icon comes up for about 3 seconds and then disappears and I’m back at my phones desktop/home screen.
When I double click the home button to check what apps are running, it seems that signal is actually running, but just with this blue logo splash screen.
Opening the app from this view brings the same issue as before, although the logo and blue colour switches for a second between slightly duller blue with a rounder speech bubble and a more saturated blue with a slightly more oval speech bubble. Maybe these are signal branding/logo changes that happened since I first installed the app and last updated the app.

**Comment by 3y3lash** [comment]
Oh - I’m really sorry, it seems I got this thread mixed up with another big report dialog I’ve been in with Signal - please disregard my last two replies!

**Comment by gnprice** [comment]
No worries :smiley:

Have you also been continuing to see symptoms in Zulip along the lines of this bug thread, or have those been resolved?

**Comment by 3y3lash** [comment]
😅
Ok, so - I just reinstalled the zulip app via App Store on the same iphone7 and the same issue is still going occurring exactly as before.

**Comment by 3y3lash** [comment]
I’m not certain if the server was upgraded on our end since then as things were pretty busy and browser version was working well so just switched to using that.

**Comment by chrisbobbe** [comment]
> I’m not certain if the server was upgraded on our end since then as things were pretty busy and browser version was working well so just switched to using that.

@3y3lash I'd recommend upgrading the Zulip server, both to see if it fixes this issue, and (unrelatedly) because there may have been several security releases since it was last upgraded, including [one from today](https://blog.zulip.com/2021/04/14/zulip-server-3-4/). 🙂

**Comment by 3y3lash** [comment]
Ooh, thanks, I’ll let them know !
